src/buf/fixed/plumbing/pool.rs

```diff
@@ -0,0 +1,186 @@
+        let raw_bufs = unsafe { ptr::NonNull::new_unchecked(iovecs.as_mut_ptr()) };
+        let orig_cap = iovecs.capacity();
+        mem::forget(iovecs);
+        Pool {
+            raw_bufs,
+            orig_cap,
+            states,
+            free_buf_head_by_cap,
+            waiting_on_next: vec![],
+        }
+    }
+
+    // If the free buffer list for this capacity is not empty, checks out the first buffer
+    // from the list and returns its data. Otherwise, returns None.
+    pub(crate) fn try_next(&mut self, cap: usize) -> Option<CheckedOutBuf> {
+        let free_head = self.free_buf_head_by_cap.get_mut(&cap)?;
+        let index = *free_head as usize;
+        let state = &mut self.states[index];
+
+        let (init_len, next) = match *state {
+            BufState::Free { init_len, next } => {
+                *state = BufState::CheckedOut;
+                (init_len, next)
+            }
+            BufState::CheckedOut => panic!("buffer is checked out"),
+        };
+
+        // Update the head of the free list for this capacity.
+        match next {
+            Some(i) => {
+                *free_head = i;
+            }
+            None => {
+                self.free_buf_head_by_cap.remove(&cap);
+            }
+        }
+
+        // Safety: the allocated array under the pointer is valid
+        // for the lifetime of self, a free buffer index is inside the array,
+        // as also asserted by the indexing operation on the states array
+        // that has the same length.
+        let iovec = unsafe { self.raw_bufs.as_ptr().add(index).read() };
+        debug_assert_eq!(iovec.iov_len, cap);
+        Some(CheckedOutBuf {
+            iovec,
+            init_len,
+            index: index as u16,
+        })
+    }
+
+    pub(crate) fn poll_next(&mut self, cap: usize, cx: &mut Context<'_>) -> Poll<CheckedOutBuf> {
+        if let Some(buf) = self.try_next(cap) {
+            return Poll::Ready(buf);
+        }
+        let waker = cx.waker();
+        if !self.waiting_on_next.iter().any(|w| w.will_wake(waker)) {
+            self.waiting_on_next.push(waker.clone());
+        }
+        Poll::Pending
+    }
+
+    fn check_in_internal(&mut self, index: u16, init_len: usize) {
+        let cap = self.iovecs()[index as usize].iov_len;
+        let state = &mut self.states[index as usize];
+        debug_assert!(
+            matches!(state, BufState::CheckedOut),
+            "the buffer must be checked out"
+        );
+
+        // Link the buffer as the new head of the free list for its capacity.
+        // Recently checked in buffers will be first to be reused,
+        // improving cache locality.
+        let next = self.free_buf_head_by_cap.insert(cap, index);
+
+        *state = BufState::Free { init_len, next };
+
+        if !self.waiting_on_next.is_empty() {
+            // Wake up tasks pending on poll_next
+            for waker in mem::take(&mut self.waiting_on_next) {
```

Comment:
Is waking them all necessary because one or more may have been canceled?

Reply:
I don't get this question, sorry. All stored wakers are used up to wake any of the tasks that still remain, and `Waker::wake` is a no-op if its task has terminated. If a task has dropped the `Next` future (or otherwise stopped calling `poll_next`), it will get a spurious wakeup once, but I think this is normal in the async world.

Reply:
Oh, you got my question. You just couldn't hear my tone of voice of my body language. Like many of my questions, it came half from naiveté and half from just wanting to make conversation and wanting to double check and keep my brain thinking during the review.

It's the first time I've seen it in action but I agree it's necessary. It's maybe a small blemish on the async world where the wake call does not return the state of whether it found a Future to put into the `Ready` queue.

Thank you.